File: 383-ransom-note/ransom-note.java
class Solution {
    public boolean canConstruct(String ransomNote, String magazine) {
        Map<Character, Integer> map = new HashMap<>();

        for(int i = 0; i < magazine.length(); i++) {
            char curChar = magazine.charAt(i);
            map.put(curChar, map.getOrDefault(curChar, 0) + 1); // char : # of Char
        }

        for(int i = 0; i < ransomNote.length(); i++) {
            char curChar = ransomNote.charAt(i);
            if(!map.containsKey(curChar) || map.get(curChar) == 0){
                return false;
            }
            map.put(curChar, map.get(curChar) - 1);
        }

        return true;
    }
}

File: 383-ransom-note/ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        hash = {}

        for c in magazine:
            hash[c] = 1 + hash.get(c, 0)
        
        for c in ransomNote:
            if c not in hash or hash[c] <= 0:
                return False
            hash[c] -= 1

        return True